### PR TITLE
Separate mutations for customer password and email change

### PR DESCRIPTION
**Pull request by sedonik**
[Customer] Allow customer password change

### Description (*)
Modify logic updateCustomer mutation next way:

Acceptance criteria:

- password field in updateCustomer must be used to set a new customer password
- Introduce new optional argument for updateCustomer called currentPassword: String. This argument must be verified on the attempt of changing email or password fields
- Existing web API tests must be adjusted and new flow for changing the password must be covered with tests

### Fixed Issues (if relevant)

1. magento/graphql-ce#917: [Customer] Allow customer password change

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/CustomerGraphQl/Model/Customer/UpdateCustomerAccount.php [hunk 1]
```diff
@@ -92,13 +102,27 @@ public function __construct(
     public function execute(CustomerInterface $customer, array $data, StoreInterface $store): void
     {
         if (isset($data['email']) && $customer->getEmail() !== $data['email']) {
-            if (!isset($data['password']) || empty($data['password'])) {
+            if (!isset($data['currentPassword']) || '' == trim($data['currentPassword'])) {
```
Comment (lenaorobei) [inline app/code/Magento/CustomerGraphQl/Model/Customer/UpdateCustomerAccount.php]: `!empty` check will be enough.

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls [hunk 2]
```diff
@@ -64,7 +64,8 @@ input CustomerInput {
     dob: String @doc(description: "The customer's date of birth")
     taxvat: String @doc(description: "The customer's Tax/VAT number (for corporate customers)")
     gender: Int @doc(description: "The customer's gender(Male - 1, Female - 2)")
-    password: String @doc(description: "The customer's password")
+    password: String @doc(description: "The new customer's password")
+    currentPassword: String @doc(description: "The current customer's password")
```
Comment (lenaorobei) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls]: Please introduce new argument in the `updateCustomer` mutation, not the new field in the `CustomerInput`.
As per ticket AC:
> Introduce new optional argument for updateCustomer called currentPassword: String

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/UpdateCustomerTest.php [hunk 3]
```diff
@@ -42,6 +42,51 @@ protected function setUp()
         $this->lockCustomer = Bootstrap::getObjectManager()->get(LockCustomer::class);
     }
 
+    /**
```
Comment (lenaorobei) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/UpdateCustomerTest.php]: Need to test also scenarios where `currentPassword` is not set and exception is thrown.

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please address review comments.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/UpdateCustomerTest.php (hunk 3 above)
Comment (sedonik) [reply dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/UpdateCustomerTest.php]: Resolved by d3e5082 commit

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls (hunk 2 above)
Comment (sedonik) [reply app/code/Magento/CustomerGraphQl/etc/schema.graphqls]: Resolved by d3e5082 commit

**Comment by sedonik** [comment]
Hi @lenaorobei 
I've updated code with review resolvings.
Could you check, please?

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please add `currentPassword: String` argument to the `updateCustomer` mutation and address builds failures.

**Comment by sedonik** [comment]
Hi @lenaorobei,
The code updated with fixes. Please check.

**Comment by lenaorobei** [comment]
Hi @sedonik! This task is currently moved to `un hold`, because the proposed approach is still under discussion.

**Comment by lenaorobei** [comment]
Hi @sedonik! Please have a look at the schema after the discussion and adjust your code. https://github.com/magento/architecture/pull/338/files
Thanks.

**Comment by chalov-anton** [comment]
Please, add a proper validation for ```date_of_birth``` and ```gender```
If perform, for example, the following GraphQL mutation 
```graphql
mutation {
  updateCustomerV2(
    input: {
      date_of_birth: "2010"
      gender: 991
    }
  ) {
    customer {
      email
      date_of_birth
      gender
    }
  }
}
```
The response is 
```graphql
{
  "data": {
    "updateCustomerV2": {
      "customer": {
        "email": "example@mail.com",
        "date_of_birth": "2010",
        "gender": 991
      }
    }
  }
}
```
In Admin, the ```date``` is set to _January 1, 1970_ The ```gender``` ie empty
gender_dateofbirth
### Expected 
- An _invalid date error_ appears for wrong the ```date``` input 
```graphql
"errors": [
    {
      "message": "Invalid date",
...
```
- The gender values are 1 for Male, 2 for Female, 3 for Not Specified.
Another input is not allowed and/or underlined in the query body.

**Comment by lenaorobei** [comment]
Closing this PR. Proper approach for updating customer will be implemented using storefront API.

**Comment by sedonik** [comment]
Hi @chalov-anton ,
Can not reproduce locally (branch "2.3-develop")
The error:
`{
  "errors": [
    {
      "message": "Cannot query field \"updateCustomerV2\" on type \"Mutation\".",
      "category": "graphql",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ]
    }
  ]
}`

**Pull request closed without merging**